### ISSUE TITLE
Why is the action obtained by sac test almost unchanged?

### ISSUE THREAD

**Issue by jaried** [aai-institute/tianshou#608]
-  I have marked all applicable categories:
-  I have visited the [source website](https://github.com/thu-ml/tianshou/)
-  I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
-  I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
0.4.7 0.21.0 1.11.0+cu113 1.20.3 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] win32
  ```
SAC is normal during training. During testing, the action is always around [high-low]/2. What is the reason?
dim=128, since my len(obs)=103, dim should not be any smaller, right?
ActorProb.forward logits values are as follows, is there a problem?
```
tensor([[0.0000, 0.1216, 0.1042, 0.0285, 0.1160, 0.0000, 0.1788, 0.1746, 0.0000,
         0.0000, 0.0000, 0.0000, 0.1407, 0.1034, 0.0000, 0.0000, 0.0429, 0.0000,
         0.0505, 0.0000, 0.0000, 0.0354, 0.0000, 0.0000, 0.0000, 0.0601, 0.0997,
         0.0000, 0.0590, 0.0349, 0.0000, 0.0431, 0.0655, 0.1495, 0.0404, 0.0126,
         0.0000, 0.0141, 0.0066, 0.0000, 0.0794, 0.0000, 0.0000, 0.0000, 0.0000,
         0.0094, 0.0979, 0.0000, 0.0000, 0.0287, 0.0709, 0.0000, 0.0744, 0.0000,
         0.0000, 0.0000, 0.0000, 0.1045, 0.0997, 0.0000, 0.0000, 0.0122, 0.0000,
         0.1931, 0.0000, 0.0000, 0.0368, 0.0000, 0.0174, 0.0000, 0.0605, 0.0000,
         0.0514, 0.0485, 0.0000, 0.0095, 0.0000, 0.1329, 0.0741, 0.0214, 0.0000,
         0.0000, 0.1195, 0.1329, 0.0139, 0.0000, 0.0516, 0.0000, 0.0803, 0.0725,
         0.0000, 0.0738, 0.0901, 0.0373, 0.0000, 0.0431, 0.0000, 0.0000, 0.1126,
         0.0000, 0.0427, 0.0000, 0.1943, 0.0000, 0.0000, 0.0000, 0.0305, 0.0000,
         0.1460, 0.0000, 0.0000, 0.0483, 0.0000, 0.0000, 0.0000, 0.0000, 0.1656,
         0.0721, 0.0000, 0.0625, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0170,
         0.0680, 0.1202]], device='cuda:0')
```
When testing, the obtained action is almost in the vicinity of (high-low)/2 ie [2.5, .5]:

```
action:[2.4866834  0.50753796]
action:[2.4825196 0.5089715]
action:[2.4895504 0.5046666]
action:[2.4803467  0.50725025]
action:[2.4762788 0.5070388]
action:[2.4858427 0.5045819]
action:[2.4855957 0.5046136]
action:[2.481365   0.50454795]
action:[2.4773862  0.50716716]
```
I've tried various methods, but the action just doesn't get the correct value.

**Comment by Trinkle23897** [comment]
I remember your action space is [0-3, 0-1]. So it seems not equal to (high-low)/2 = [1.5, .5].

Have you tried an even smaller network? For mujoco task, two 256 hidden layers are actually enough
https://github.com/thu-ml/tianshou/blob/5eab7dc2180a2f3915943c439c173b4520bd829e/examples/mujoco/mujoco_sac.py#L27

**Comment by jaried** [comment]
My action space has been changed to `self.action_space = spaces.Box(low=np.array([0, 0]), high=np.array([5, 1]))` ,
hidden_size is changed to `parser.add_argument('--hidden-sizes', type=int, nargs='*', default=[128, 128])`.
Observation_space is changed to `self.observation_space = spaces.Box(low=-1, high=1, shape=(103,), dtype=np.float32)`.

But the resulting action is still [high-low]/2:
action:[2.515202   0.49850938]
action:[2.5156512 0.4981308]
action:[2.4791143  0.50075144]
action:[2.462799   0.50011784]
action:[2.4861367 0.504844 ]
action:[2.4674678 0.5043308]
action:[2.492096   0.50269437]
action:[2.4920223  0.50268203]
action:[2.521101   0.50354785]
action:[2.5050008 0.4998747]
action:[2.4863014 0.5048039]
action:[2.4939923  0.50133646]
action:[2.5046067 0.5013162]
action:[2.5046992 0.5013042]
action:[2.490491   0.50049096]
action:[2.5144956 0.5004514]
![1650839055(1)](https://user-images.githubusercontent.com/26672425/164999218-f716b5d3-f5b9-4d4e-b5f3-0cad61df3bb8.png)

**Comment by Trinkle23897** [comment]
![2022-04-24 20-50-41 的屏幕截图](https://user-images.githubusercontent.com/8189182/165004316-b49cf14f-0350-461a-90e8-9e1b174a3af5.png)

I suspect the agent went to the local minimum. It's quite hard to say what's wrong...
~Also, I don't think an agent can learn something from only 70k iterations...~

**Comment by jaried** [comment]
This is just the number of steps to learn, I used 20 parallel environments, so env_steps is multiplied by 20, and is about 1M.

What else can I check?
Judging from the results, sac has learned something, but the test will not stay near [high-low]/2, right?

![1650848210(1)](https://user-images.githubusercontent.com/26672425/165004646-9489d89a-2258-48a0-85ed-92cb6da460bd.png)

**Comment by Trinkle23897** [comment]
Did you evaluate the model at about 10k/70k early stage?

**Comment by jaried** [comment]
I don't understand how the evaluation model you are referring to does. 

The training has yielded good results, but in my current test, the result of the action is still:
```
action:[2.5345519  0.50954944]
action:[2.5337598 0.5090791]
action:[2.5364976 0.5098061]
action:[2.517208   0.50975025]
action:[2.5180233 0.5098032]
action:[2.5173464 0.5096032]
action:[2.5271697 0.5083998]
action:[2.5410113  0.50793535]
action:[2.5316892  0.50925136]
action:[2.5315988  0.50919783]
```

**Comment by Trinkle23897** [comment]
Mark as resolved. The issue is, the "normal" actions are from `rsample` instead of `argmax`.

**Comment by jaried** [comment]
I used the network training of [256,256], until the log coordinates, when the loss of critic did not decrease significantly, I still did not get a better training effect. I get much better training results with a network of [512,512,512,512].

**Comment by jaried** [comment]
Is it possible to get `argmax` with `act = logits[0]`? It feels that the `act` obtained with `act = dist.rsample()` is not stable.

**Comment by Trinkle23897** [comment]
What do you mean with argmax?

**Comment by jaried** [comment]
Can `act = logits[0]` be modified? I feel that after adding `deterministic_eval=False` to `SACPolicy`, the `action` obtained by `act = dist.rsample()` is not stable.
